staker/multi_protocol_staker.go

```diff
@@ -0,0 +1,207 @@
+		panic("RollupCore ABI missing AssertionCreated event")
+	}
+	assertionCreatedId = assertionCreatedEvent.ID
+}
+
+type MultiProtocolStaker struct {
+	stopwaiter.StopWaiter
+	bridge     *bridgegen.IBridge
+	oldStaker  *Staker
+	boldStaker *BOLDStaker
+}
+
+func NewMultiProtocolStaker(
+	l1Reader *headerreader.HeaderReader,
+	wallet ValidatorWalletInterface,
+	callOpts bind.CallOpts,
+	config L1ValidatorConfigFetcher,
+	blockValidator *BlockValidator,
+	statelessBlockValidator *StatelessBlockValidator,
+	stakedNotifiers []LatestStakedNotifier,
+	confirmedNotifiers []LatestConfirmedNotifier,
+	validatorUtilsAddress common.Address,
+	bridgeAddress common.Address,
+	fatalErr chan<- error,
+) (*MultiProtocolStaker, error) {
+	oldStaker, err := NewStaker(
+		l1Reader,
+		wallet,
+		callOpts,
+		config,
+		blockValidator,
+		statelessBlockValidator,
+		stakedNotifiers,
+		confirmedNotifiers,
+		validatorUtilsAddress,
+		fatalErr,
+	)
+	if err != nil {
+		return nil, err
+	}
+	bridge, err := bridgegen.NewIBridge(bridgeAddress, oldStaker.client)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiProtocolStaker{
+		oldStaker:  oldStaker,
+		boldStaker: nil,
+		bridge:     bridge,
+	}, nil
+}
+
+func (m *MultiProtocolStaker) IsWhitelisted(ctx context.Context) (bool, error) {
+	return m.oldStaker.IsWhitelisted(ctx)
+}
+
+func (m *MultiProtocolStaker) Initialize(ctx context.Context) error {
+	boldActive, rollupAddress, err := m.isBoldActive(ctx)
+	if err != nil {
+		return err
+	}
+	if boldActive {
+		log.Info("BOLD protocol is active, initializing BOLD staker")
+		boldStaker, err := m.setupBoldStaker(ctx, rollupAddress)
+		if err != nil {
+			return err
+		}
+		m.boldStaker = boldStaker
+		return m.boldStaker.Initialize(ctx)
+	}
+	log.Info("BOLD protocol not detected on startup, using old staker until upgrade")
+	return m.oldStaker.Initialize(ctx)
+}
+
+func (m *MultiProtocolStaker) Start(ctxIn context.Context) {
+	m.StopWaiter.Start(ctxIn, m)
+	if m.oldStaker.Strategy() != WatchtowerStrategy {
+		m.oldStaker.wallet.Start(ctxIn)
+	}
+	if m.boldStaker != nil {
+		log.Info("Starting BOLD staker")
+		m.boldStaker.Start(ctxIn)
+		m.StopOnly()
+	} else {
+		log.Info("Starting pre-BOLD staker")
+		m.oldStaker.Start(ctxIn)
+		stakerSwitchInterval := m.oldStaker.config().BOLD.CheckStakerSwitchInterval
+		m.CallIteratively(func(ctx context.Context) time.Duration {
+			switchedToBoldProtocol, err := m.checkAndSwitchToBoldStaker(ctxIn)
+			if err != nil {
+				log.Error("staker: error in checking switch to bold staker", "err", err)
+				return stakerSwitchInterval
+			}
+			if switchedToBoldProtocol {
+				log.Info("Detected BOLD protocol upgrade, stopping old staker and starting BOLD staker")
+				// Ready to stop the old staker.
+				m.oldStaker.StopOnly()
+				m.StopOnly()
+			}
+			return stakerSwitchInterval
+		})
+	}
+}
+
+func (m *MultiProtocolStaker) StopAndWait() {
+	if m.boldStaker != nil {
+		m.boldStaker.StopAndWait()
+	}
+	m.oldStaker.StopAndWait()
```

Comment:
that will panic if start wasn't called for oldStaker, which seems to be the case if we started with when BOLD is active

Reply:
One solution might be something like:
If on init you see bold is active, keep oldStaker = nil - and here check if it's nil before stopAndWait

Reply:
Done.